test/integration/hds_integration_test.cc

```diff
@@ -411,17 +408,13 @@ TEST_P(HdsIntegrationTest, SingleEndpointHealthyTcp) {
   RELEASE_ASSERT(result, result.message());
   host_upstream_->set_allow_unexpected_disconnects(true);
 
-  // Make sure the cluster processed the endpoint's response
-  test_server_->waitForCounterGe("cluster.anna.health_check.success", 1);
-
-  // Envoy reports back to server
+  // Receive updates until the one we expect arrives
   ASSERT_TRUE(hds_stream_->waitForGrpcMessage(*dispatcher_, response_));
-
-  // Check that the response is correct
-  auto endpoint = response_.endpoint_health_response().endpoints_health(0);
-  EXPECT_EQ(envoy::api::v2::core::HealthStatus::HEALTHY, endpoint.health_status());
-  EXPECT_EQ(host_upstream_->localAddress()->ip()->port(),
-            endpoint.endpoint().address().socket_address().port_value());
+  while (!checkEndpointHealthResponse(response_.endpoint_health_response().endpoints_health(0),
+                                      envoy::api::v2::core::HealthStatus::HEALTHY,
+                                      host_upstream_->localAddress())) {
+    ASSERT_TRUE(hds_stream_->waitForGrpcMessage(*dispatcher_, response_));
```

Comment:
This `while` pattern appears a lot, can you factor it into a `waitForEndpointHealthResponse` method? You can skip doing this for the more complex case of waiting for two conditions further below, but I think most of the cases are simple.